src/main/java/org/truffleruby/core/rope/NativeRope.java

```diff
@@ -20,8 +20,8 @@
 
     private Pointer pointer;
 
-    public NativeRope(FinalizationService finalizationService, byte[] bytes, Encoding encoding, int characterLength) {
-        super(encoding, CodeRange.CR_UNKNOWN, false, bytes.length, characterLength, 1, null);
+    public NativeRope(FinalizationService finalizationService, byte[] bytes, Encoding encoding, int characterLength, CodeRange codeRange) {
+        super(encoding, codeRange, false, bytes.length, characterLength, 1, null);
```

Comment:
That CodeRange might become invalid later on since bytes can be written to the native memory directly.
cc @nirvdrum

Reply:
My understanding is it's undefined behavior if an extension modifies an `RString`'s data after MRI is handed back control of the `RString`. In that case, we should be okay caching the code range. Even MRI caches the code range for strings -- they just do it lazily. But, if we truly need to worry about native code modifying a string after the code range has been set, we'll need to change up a lot of things. Our guards and specializations simply don't handle propagation of `CR_UNKNOWN`.

Reply:
Interesting, it seems indeed MRI keeps the code range even after `RSTRING_PTR`:
```ruby
s = "abc";
p s.valid_encoding?; # true
p modify_rstring s # RSTRING_PTR(s)[1] = 0xFF; in C : "a\xFFc"
p s.valid_encoding?; # true but wrong!
s[1]="\xFF"
p s.valid_encoding? # false, now it noticed it
```

So I guess the responsibility to update the code range if needed falls to the C ext programmer in such a case (with `ENC_CODERANGE_SET` and co I guess).

Reply:
I think we no-op on `ENC_CODERANGE_SET`. We may want to add an assertion there at the very least. We could make the code range mutable, too, but we'll want to add a new node for that then.

Reply:
It seems we don't define ENC_CODERANGE_SET at all currently.
But indeed once added it should faithfully implement the semantics or error out.

Reply:
You're right. I was thinking of some work @bjfish is doing. We should have `ENC_CODERANGE_SET` soon.